async/await plus a variety of other changes

ggriffiniii opened the pull request:
I'm sending this at the risk of overwhelming as I recognize it's a pretty large pull request. This started with moving to std::future::Future as discussed in #111 then snowballed into a variety of changes (hopefully improvements) that could be made while we were already breaking semver compatibility.

Feel free to pick and choose which commits you like. The first commit is the only one necessary to get to std::future::Futures. As mentioned in the feature request this currently pins a few crates that are released with alpha labels. I would expect to wait for those dependencies to no longer be in alpha before releasing these changes.

ggriffiniii committed:
Move to std::futures to support async/await.

ggriffiniii committed:
Require trait implementations to be Send + Sync.

Tidy up some of the trait bounds on types and methods.

ggriffiniii committed:
FlowType isn't used for anything. Remove it.

ggriffiniii committed:
Have TokenStorage take scopes by iterator rather than Vec.

This reduces the number of allocations needed.

ggriffiniii committed:
Accept scopes as a slice of anything that can produce a &str.

Along with the public facing change the implementation has been modified
to no longer clone the scopes instead using the pointer to the scopes
the user provided. This greatly reduces the number of allocations on
each token() call.

Note that this also changes the hashing method used for token storage in
an incompatible way with the previous implementation. The previous
implementation pre-sorted the vector and hashed the contents to make the
result independent of the ordering of the scopes. Instead we now combine
the hash values of each scope together with XOR, thus producing a hash
value that does not depend on order without needing to allocate another
vector and sort.

ggriffiniii committed:
Remove instances of cloning ApplicationSecret

ApplicationSecret is not a small struct. This removes the instances
where it's cloned in favor of passing a shared reference.

ggriffiniii committed:
Authenticator.store no longer needs to be reference counted.

ggriffiniii committed:
No need to clone the hyper::Client

The ownership behavior is straightforward and more clear when not
cloning arbitrary handles.

ggriffiniii committed:
The inner GetToken on Authenticator no longer needs to be reference counted.

ggriffiniii committed:
Make some of the helpers a bit more idiomatic.

ggriffiniii committed:
Some more improvements to reduce unnecessary allocations.

ggriffiniii committed:
Add JsonErrorOr enum to make json error handling  more concise/consistent.

JsonErrorOr is an untagged enum that is generic over arbitrary data.
This means that when deserializing JsonErrorOr<T> it will first check
the json field for an 'error' attribute. If one exists it will
deserialize into the JsonErrorOr::Err variant that contains a JsonError.
If the message doesn't contain an 'error' field it will attempt to
deserialize T into he JsonErrorOr::Data variant.

ggriffiniii committed:
cargo fmt

ggriffiniii committed:
cargo clippy fixes

ggriffiniii committed:
Remove the HTTPRedirectEphemeral variant.

In favor of making it the default and removing the option to specify a
port to listen on. If needed a variant can be added to specify a port
explicitly, but most users should want an ephemeral port chosen so
making it the default makes sense while other breaking changes are in
flight.

ggriffiniii committed:
Remove unnecessary trait bounds on hyper connector.

Send+Sync is implied by the trait, and Clone is no longer necessary.

ggriffiniii commented on Cargo.toml:
```diff
@@ -14,19 +14,18 @@ edition = "2018"
-tokio-threadpool = "0.1"
-tokio = "0.1"
-tokio-timer = "0.2"
+futures-preview = "=0.3.0-alpha.19"
```
We need to align the dependency versions with the versions hyper is using so that traits match. Once hyper releases a new version we can update accordingly.

ctaggart replied:
https://github.com/hyperium/hyper/pull/2007 is the PR for hyper

ggriffiniii committed:
Delegates no longer need to implement Clone.

ggriffiniii committed:
Unify trait bounds on Authenticator::build

ggriffiniii committed:
Remove more unnecessary clones.

ggriffiniii committed:
Refactor retrieve_device_token.

wait_for_device_token polls indefinitely at the specified intervals.
Use tokio::timer::Timeout to bound the time it will poll for.

dermesser commented:
btw, I know that you are working on this (and am very happy about it!), I just currently lack the time to help more :)

ggriffiniii committed:
Remove unnecessary 'static bounds

ggriffiniii committed:
Refactor JWT handling in ServiceAccountAccess.

Avoid reading and parsing the private key file on every invocation of
token() in favor or reading it once when the ServiceAccountAccess is
built. Also avoid unnecessary allocations when signing JWT tokens and
renamed sub to subject to avoid any confusion with the std::ops::Sub
trait.

ggriffiniii committed:
Authenticator should handle the server not returning a refresh_token.

Currently the authenticator will panic when trying to refresh an expired
token that does not have a refresh token. This change handles it so that
the authenticator will only attempt a refresh when a refresh_token
exists, and otherwise will attempt to retrieve a fresh token.

ggriffiniii committed:
Refactor token storage.

The current code uses standard blocking i/o operations (std::fs::*) this
is problematic as it would block the entire futures executor waiting for
i/o.

This change is a major refactoring to make the token storage mechansim
async i/o friendly. The first major decision was to abandon the GetToken
trait. The trait is only implemented internally and there was no
mechanism for users to provide their own, but async fn's are not
currently supported in trait impls so keeping the trait would have
required Boxing futures. This probably would have been fine, but seemed
unnecessary. Instead of a trait the storage mechanism is just an enum
with a choice between Memory and Disk storage.

The DiskStorage works primarily as it did before, rewriting the entire
contents of the file on every set() invocation. The only difference is
that we now defer the actual writing to a separate task so that it does
not block the return of the Token to the user. If disk i/o is too slow
to keep up with the rate of incoming writes it will push back and
will eventually block the return of tokens, this is to prevent a buildup
of in-flight requests. One major drawback to this approach is that any
errors that happen on write are simply logged and no delegate function
is invoked on error because the delegate no longer has the ability to
say to sleep, retry, etc.

ggriffiniii committed:
allow setting grant_type for device code

ggriffiniii committed:
Make FlowDelegate object safe.

ggriffiniii committed:
Major refactor of the public API.

1) Remove the GetToken trait. The trait seemed to be organically
designed. It appeared to be mostly tailored for simplifying the
implementation since there was no way for users to provide their own
implementation to Authenticator. It sadly seemed to get in the way of
implementations more than it helped. An enum representing the known
implementations is a more straightforward way to accomplish the goal and
also has the benefit of not requiring Boxing when returning features
(which admittedly is a minor concern for this use case).

2) Reduce the number of type parameters by using trait object for
delegates. This simplifies the code considerably and the performance
impact of virtual dispatch for the delegate calls is a non-factor.

3) With the above two simplifications it became easier to unify the
public interface for building an authenticator. See the examples for how
InstalledFlow, DeviceFlow, and ServiceAccount authenticators are now created.

ggriffiniii committed:
Minimize the number of items on the rustdoc landing page.

Restructure the modules and imports to increase the signal to noise
ration on the cargo doc landing page. This includes exposing some
modules as public so that they can contain things that need to be public
but that users will rarely need to interact with. Most items from
types.rs were moved into an error.rs module that is now exposed
publicly.

ggriffiniii committed:
Rename RequestError to Error

RequestError is the error value that encompasses all errors from the
authenticators. Their is an established convention of using Error as the
name for those types.

ggriffiniii committed:
Tidy up some imports.

No more need to macro_use serde. Order the imports consistently (albeit
somewhat arbitrary), starting with items from this crate, followed by
std, followed by external crates.

ggriffiniii committed:
Improve documentation

ggriffiniii committed:
Tidy up tests.

ggriffiniii committed:
Remove an obsolete todo.

ggriffiniii committed:
Remove dev-dependencies that are no longer used.

These appear to only be used by examples in the old/ directory which has
not compiled for a long time. Not sure why the contents of that
directory are still around.

ggriffiniii committed:
Add header styling to the AuthenticatorBuilder rustdoc

ggriffiniii committed:
refactor storage get and set methods.

These previously accepted a hash and scopes. The hash was required to be
a hash of the provided scopes but that wasn't enforced by the compiler.
We now have the compiler enforce that by creating a HashedScopes type
that ties the scopes and the hash together and pass that into the
storage methods.

ggriffiniii committed:
storage set method should just accept a Token rather than Option<Token>.

No caller ever provided a None value. Presumably a None value should
delete the token, but it didn't do that and that would be more clearly
done with a remove or delete method.

ggriffiniii committed:
JSONToken should always contain scopes.

This is already the case when writing a token file. Presumably the only
reason it was an Option was for backwards compatibility, but we're
already breaking compatibility with the change to the hash value so this
seems like an appropriate time to make the change.

This change also highlights how unused the hash value has been
previously. Future changes plan to use the hash value for more efficient
handling.

ggriffiniii committed:
Use seahash rather that DefaultHasher.

DefaultHasher is not documented as being consistent. It's best to not
trust that the resulting hash value is consistent even across different
executions of the same binary and even more so across different
versions.

ggriffiniii committed:
Use the storage token hash more effectively.

Use a BTreeMap to key the tokens by the hash value. On retrieval first
look for a matching hash value and return it if it exists. Only if it
does not exist does it fallback to the subset matching. This makes the
common case where an application uses a consistent set of scopes more
efficient without detrimentally impacting the less common cases.

ggriffiniii committed:
Don't serialize the scope hash.

Seahash is a stable hash, but there isn't any value in serializing it's
value. Instead calculate the value of the hash when deserializing and
only serialize the scopes and tokens. This provides flexibility to
change the hash value in the future without breaking the on-disk format.

ggriffiniii committed:
Use a bloom filter to track scopes.

Each token is stored along with a 64bit bloom filter that is created
from the set of scopes associated with that token. When retrieving
tokens for a set of scopes a new bloom filter is calculated for the
requested scopes and compared to the filters of all previously fetched
scopes. The bloom filter allows for efficiently skipping entries that
are definitely not a superset.

ctaggart commented:
These changes look great @ggriffiniii. @dermesser, any plans to review & merge, or add @ggriffiniii as a collaborator? I upgraded a command line application of ours to rust async/.await and now I'm going to see if I can use this to get device auth working with Azure. It would be nice to see it land on `master`.

ggriffiniii commented:
> These changes look great @ggriffiniii. @dermesser, any plans to review & merge, or add @ggriffiniii as a collaborator? I upgraded a command line application of ours to rust async/.await and now I'm going to see if I can use this to get device auth working with Azure. It would be nice to see it land on `master`.

Let me know if you run into any issues.

ctaggart commented:
@ggriffiniii, the format of `tokenstorage.json` has changed. With or without the file, I get:
```
thread 'main' panicked at 'authenticator: Custom { kind: InvalidData, error: Error("invalid type: map, expected a sequence", line: 1, column: 0) }', src/libcore/result.rs:1187:5
```

I improved the error so that it says:
```
     Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
thread 'main' panicked at 'authenticator: Custom { kind: InvalidData, error: "loading file tokenstorage.json: invalid type: map, expected a sequence at line 1 column 0" }', src/libcore/result.rs:1187:5
```

In storage.rs:
```
        let token_vec: Vec<JSONToken> = serde_json::from_slice(&contents)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("loading file {}: {}", filename.display(), e)))?;
```

I'm working on a branch that merged your branch in. I'll create a new one and try again just to make sure.

ggriffiniii commented:
Yes, my branch changes the on disk format of tokenstorage.json. I'm assuming you're seeing that error when attempting to read a file cretaed with an old version? Do you still get the error if you delete the old file and start fresh? I should have called that out more explicitly in the pull request summary.

ctaggart commented:
On this branch with a clientsecret.json, but a tokenstorage.json removed, I get:
```
cameron@d5520:~/github/yup-oauth2/examples/test-device$ cargo run
   Compiling yup-oauth2 v3.1.1 (/home/cameron/github/yup-oauth2)
   Compiling test-device v0.1.0 (/home/cameron/github/yup-oauth2/examples/test-device)
    Finished dev [unoptimized + debuginfo] target(s) in 7.03s
     Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
thread 'main' panicked at 'authenticator: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1187:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

ggriffiniii committed:
Fix a bug in refactoring the storage layer.

Attempting to load from disk when the file does not exist should not
return an error and should continue with an empty set of tokens.

ggriffiniii commented:
> On this branch with a clientsecret.json, but a tokenstorage.json removed, I get:
> 
> ```
> cameron@d5520:~/github/yup-oauth2/examples/test-device$ cargo run
>    Compiling yup-oauth2 v3.1.1 (/home/cameron/github/yup-oauth2)
>    Compiling test-device v0.1.0 (/home/cameron/github/yup-oauth2/examples/test-device)
>     Finished dev [unoptimized + debuginfo] target(s) in 7.03s
>      Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
> thread 'main' panicked at 'authenticator: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1187:5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
> ```

Oh, that was a dumb mistake when refactoring. Pushed an update that should hopefully fix it.

ctaggart commented:
Thank you! I confirmed it works with both Google and Azure.

ggriffiniii committed:
Tie ServiceAccount's into Authenticator.

Prior to this change DeviceFlow and InstalledFlow were used within
Authenticator, while ServiceAccountAccess was used on it's own. AFAICT
this was the case because ServiceAccountAccess never used refresh tokens
and Authenticator assumed all tokens contained refresh tokens.
Authenticator was recently modified to handle the case where a token
does not contain a refresh token so I don't see any reason to keep the
service account access separate anymore. Folding it into the
authenticator provides a nice consistent interface, and the service
account implementation no longer needs to provide it's own caching since
it is now handled by Authenticator.

ggriffiniii committed:
Rename service_account_key_from_file to read_service_account_key

This makes the name consistent with the other helper
read_application_secret.

ggriffiniii committed:
Remove what appears to be an unnecessary replace.

decode_rsa_key does a .replace('\\n', '\n') which replaces any literal
2 byte sequence '\n' with a newline character. The original commit that
added it is 38fd8514933bedaa5b23d7ba0b02c35353e3f05b but there's no
mention of why it's needed and there are no test cases that fail when
it's omitted. Any file that has a literal 2 byte sequence of '\\n' is
surely not a valid pkcs8 file so it seems perfectly valid to skip the
replace (and allocation) and return an error if one is encountered.

If it's determined that this check is needed for some reason please add
a unit test that explains why for future contributors.

ggriffiniii committed:
Remove AuthenticatorDelegate::{client_error, request_failure}

These methods are never called.

ggriffiniii committed:
Make helpers that read from disk async

ggriffiniii committed:
Split FlowDelegate into DeviceFlowDelegate and InstalledFlowDelegate.

Each flow invokes a non-overlapping set of methods. There doesn't appear
to be any benefit in having both flows use a common trait. The benefit
of splitting the traits is that it makes it clear which methods need to
be updated for each flow type where previously comments were required to
communicate that information.

ggriffiniii committed:
Move the impls for PollError into error.rs

These should have been moved when error.rs was created but were missed.

ggriffiniii committed:
Reconsider Error variants

BadServerResponse didn't seem to be adequately different from
NegativeServerResponse so it's been removed. NegativeServerResponse is
now a struct variant with field names 'error' and 'error_description' to
self document what it contains.

InvalidClient and InvalidScope were only ever created based on string
parsing of the returned error message from the server. This seemed
overly specific to a particular implementation and didn't provide much
value to callers. Printing a NegativeServerResponse would provide users
the same information.

The caching layer never returns errors anymore so remove that variant.

ggriffiniii committed:
InstalledFlowDelegate::present_user_url should return a String error.

Prior to this change the only place present_user_url is called overwrote
the error with a static string. After this change the error returned is
appended to the message. No need to make the signature more complicated
when the error is always going to be flattened to a string anyway.

ggriffiniii committed:
Remove code to strip trailing newlines for backwards compatibility.

Based on the comment in the code the justification for the change was
because old FlowDelegates used to contain newlines and changing how the
returned string from the delegate was handled would be a breaking
change. In this case it should be safe to remove the hack because we're
breaking compatibility. All users that once implemented FlowDelegate
will now need to implement InstalledFlowDelegate and uphold the new
contract which implicitly means the authcode returned should represent
the authcode and nothing more. No manipulation of the returned string
will be done.

ggriffiniii committed:
Rename Error::ClientError and RefreshError::ConnectionError to HttpError.

PollError already contained an HttpError variant so this makes all
variants that contain a hyper::Error consistently named.

ggriffiniii committed:
Refactor error handling and as a consequence delegates.

This Removes RefreshError and PollError. Both those types can be fully
represented within Error and there seems little value in distinguishing
that they were resulting from device polling or refreshes. In either
case the user will need to handle the response from token() calls
similarly. This also removes the AuthenticatorDelegate since it only
served to notify users when refreshes failed, which can already be done
by looking at the return code from token. DeviceFlow no longer has the
ability to set a wait_timeout. This is trivial to do by wrapping the
token() call in a tokio::Timeout future so there's little benefit for
users specifying this value. The DeviceFlowDelegate also no longer has
the ability to specify when to abort, or alter the interval polling
happens on, but it does gain understanding of the 'slow_down' response
as documented in the oauth rfc. It seemed very unlikely the delegate was
going to do anything other that timeout after a given time and that's
already possible using tokio::Timeout so it needlessly complicated the
implementation.

ggriffiniii committed:
Improve Token

Remove expires_in in favor of only having an expires_at DateTime field.
Add a from_json method that deserializes from json data into the
appropriate Token (or Error) and use that consistently throughout the
codebase.

ggriffiniii committed:
Rename PollInformation DeviceAuthResponse.

Have it correctly handle either verification_uri or verification_url and
deserialize into a struct that has the data types desired.

ggriffiniii committed:
Make DeviceFlowDelegate::present_user_code return a Future.

This is to allow for implementations to use async code. The returned
Future will be awaited before polling for the token begins.

ggriffiniii committed:
Fix a bug introduced in the storage layer.

When bloom filters were added the btreemap values changed to be a
vector of tokens to accomodate the possibility of bloom filter
collisions. The implementation naively just pushed new tokens onto the
vec even if they were replacing previous tokens meaning old tokens were
still kept around even after a refresh has replaced it. To fix this
efficiently the storage layer now tracks both a hash value and a bloom
filter along with each token. Their is a map keyed by hash for every
token that points to a reference counted version of the token, and each
token also exists in a separate vector. Updates to existing tokens
happens in place, when new entries are added they are added to both data
structures.

ghost commented:
Hi @ggriffiniii Thanks for your superb work :-),  I was using this pull request for Google STT it was working great till [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c). After latest update, I am getting a minor error regarding `JSONToken` 

```rust
   async fn stream(&self, uri: Uri) {
    |  ______________________^
   | |         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)
   | |             .await
   | |             .expect("Failed to read service account file");
...  |
   | |         }
   | |     }
   | |_____^ `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>` cannot be sent between threads safely
    |
```
    = help: the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>`

ggriffiniii commented:
> Hi @ggriffiniii Thanks for your superb work :-), I was using this pull request for Google STT it was working great till [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c). After latest update, I am getting a minor error regarding `JSONToken`
> 
> ```rust
>    async fn stream(&self, uri: Uri) {
>     |  ______________________^
>    | |         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)
>    | |             .await
>    | |             .expect("Failed to read service account file");
> ...  |
>    | |         }
>    | |     }
>    | |_____^ `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>` cannot be sent between threads safely
>     |
> ```
> 
> ```
> = help: the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>`
> ```

Thanks for the report. Are you sure it was commit d291ebb? The reference to Rc<RefCell<JSONToken>> makes me suspect the most recent change 9dd49d4 would be the culprit. Do you have any more context around the code that's failing to compile. I'll take a look and see if I can figure out what's going on, but it's not clear from that snippet why the JSONToken needs to be send.

BTW. Thanks for testing these changes. I appreciate it.

ggriffiniii committed:
Use Arc<Mutex<T>> rather than Rc<RefCell<T>> in DiskStorage.

This keeps DiskStorage Sync + Send and therefore Authenticator Sync +
Send. The DiskStorage was threadsafe because JSONTokens contains a Mutex
around all the Rc<RefCell<T>> objects, but there's no way to prove to
the type system that none of the Rc's get cloned to an alias used
outside the Mutex so it's not provably safe. I'll probably reevaluate
the design here, but in the meantime the double locking is fine.

ggriffiniii commented:
Ok. I think b3b8b73 will fix your issue. I'll add some tests to ensure this doesn't happen again.

ggriffiniii committed:
cargo fmt

ggriffiniii committed:
Add a test to ensure that Authenticator is Send+Sync

ggriffiniii committed:
Refactor storage to only use a BTreeMap.

Keeping the same tokens in a Vec and BTreeMap created more overhead than
was warranted. It makes much more sense to simply iterator over the
BTreeMap than keep a separate Vec.

ghost commented:
@ggriffiniii Thanks for looking into it and fixing the issue.

> Are you sure it was commit d291ebb? The reference to Rc<RefCell> makes me suspect the most recent change 9dd49d4 would be the culprit.

Actually, I meant to say that when I first pulled the source, latest revision was at [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c), which was working fine. However, when I fetched again, the culprit code which you pointed out was somewhere in the last 10 commits which I hadn't looked into.

 I tested after your fix it is working now :-), Thanks.

ghost commented:
One question, does it refresh service account token automatically once expired?

ggriffiniii commented:
> One question, does it refresh service account token automatically once expired?
I believe it should.

ggriffiniii committed:
Go back to waiting for disk writes on every token set.

Defering disk writes is still probably a good idea, but unfortunately
there are some tradeoffs with rust's async story that make it non-ideal.
Ideally we would defer writes, but have a Drop impl on DiskStorage that
waited for all the deferred writes to complete. While it's trival to
create a future that waits for all deferred writes to finish it's not
currently possible to write a Drop impl that waits on a future.

It would be possible to write an inherent async fn that takes self by
value and waits for the writes, but that method would need to be
propogated up all the way to users of the library and they would need to
remember to invoke it before dropping the Authenticator.

ggriffiniii committed:
Move all the end to end tests into an integration test

All the same functionality can be tested through the publicly exposed
API providing more extensive coverage.

ghost commented:
@ggriffiniii I used a following method to retrieve access token:

```rust
    async fn get_access_token(&self) -> String {    
        let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)    
            .await    
            .expect("Failed to read service account file");    
    
        let auth = ServiceAccountAuthenticator::builder(secret)    
            .build()    
            .await    
            .expect("Invalid credentials");    
    
        let scopes = &[&self.credentials.scope];    
        let token = auth.token(scopes).await.expect("Couldn't get token");    
        token.access_token    
    } 
```
If I use this immediately it works fine. However, when I send the request to the server may be after 2 or 3 hours, I get an error from google for invalid token, I didn't close the application through out the session. So apparently it seems the token is not refreshing automatically. I may be missing something here. The token string should remain same if token is refreshed just before expiry, right? Also is refreshing of token behind any timer?

ggriffiniii commented:
> I used a following method to retrieve access token:
> 
> ```rust
>     async fn get_access_token(&self) -> String {    
>         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)    
>             .await    
>             .expect("Failed to read service account file");    
>     
>         let auth = ServiceAccountAuthenticator::builder(secret)    
>             .build()    
>             .await    
>             .expect("Invalid credentials");    
>     
>         let scopes = &[&self.credentials.scope];    
>         let token = auth.token(scopes).await.expect("Couldn't get token");    
>         token.access_token    
>     } 
> ```
> 
> If I use this immediately it works fine. However, when I send the request to the server may be after 2 or 3 hours, I get an error from google for invalid token, I didn't close the application through out the session. So apparently it seems the token is not refreshing automatically. I may be missing something here. The token string should remain same if token is refreshed just before expiry, right? Also is refreshing of token behind any timer?

@abdul-rehman0 Ok. I think there's a misunderstanding. Your code creates an authenticator, retrieves a token, and returns the access_token string (destroying the authenticator once the function returns). This is probably not what you what you want. The access token that is returned is typically only valid for a relatively short lifetime (the 2-3 hours that you experience). The authenticator is responsible for refreshing an access token but will only do so when the `token` method is invoked. When you invoke the `token` method it first checks to see if it has retrieved any tokens that are still valid for the scopes requested, if so it will return the cached access token. If the cached access token is expired (or about to expire) it will retrieve a new access token.

To summarize: you should create a single long-lived authenticator and whenever you need an access token you invoke the `token` method and use it. By doing this you'll ensure that the access token is valid.

Let me know if that makes sense and if that solves the expiration problem you're having.

dermesser commented:
I am following along a bit here - it seems you are still productively working on it, but please let me know once you have it in a form where I can look to merge it! :)

ggriffiniii committed:
Authenticator now returns an AccessToken.

What was previously called Token is now TokenInfo and is merely an
internal implementation detail. The publicly visible type is now called
AccessToken and differs from TokenInfo by not including the refresh
token. This makes it a smaller type for users to pass around as well as
reducing the ways that a refresh token may be leaked. Since the
Authenticator is responsible for refreshing the tokens there isn't any
reason users should need to concern themselves with refresh tokens.

ggriffiniii committed:
Extend the refresh tests.

Verify that a second refresh can happen after the first. This adds
coverage to ensure that a refresh flow keeps the refresh token intact by
showing that a second refresh can succeed.

ggriffiniii commented:
> I am following along a bit here - it seems you are still productively working on it, but please let me know once you have it in a form where I can look to merge it! :)

Great to hear. I've been continually hacking away and the changes have definitely become more and more opinionated the more I've done, so don't hesitate to push back on things. In general my goal is to keep the tip of this branch as a ready-to-merge candidate, with the exception that the dependencies are still alpha releases. tokio recently released a real 0.2 release, so the only dependency we're still waiting for is hyper. Once that happens I'll ping this thread and say it's ready for a look.

ggriffiniii committed:
Add debug logging.

Could be helpful when troubleshooting issues with various providers if
the user is able to turn on debug logging. The most critical logging
provided is the request and responses sent and received from the oauth
servers.

ghost commented:
> To summarize: you should create a single long-lived authenticator and whenever you need an access token you invoke the token method and use it. By doing this you'll ensure that the access token is valid.

> Let me know if that makes sense and if that solves the expiration problem you're having.

@ggriffiniii thanks for the clarification, this approach works for me I tested it :).

jroesch commented:
I'm excited for this release as the version update here and Google APIs means we can drop the tokio 1.0 dependence and upgrade multiple services at work. It looks like hyper is releasing tonight: https://github.com/hyperium/hyper/pull/2061 (or real soon).

ggriffiniii commented:
hyper 0.13 is released! Now the next hurdle is getting a hyper-rustls release that supports hyper 0.13 :(

There is a PR already in the works here: https://github.com/ctz/hyper-rustls/pull/96

ctaggart commented:
`persist_tokens_to_disk` is saving the tokens to a file, but the file should not be readable to other users. Permissions are currently 644, but should be 600. Should I move this to a separate issue?

``` rust
    let authenticator = DeviceFlowAuthenticator::builder(app_secret)
        .persist_tokens_to_disk(token_path)
        .device_code_url(format!("https://login.microsoftonline.com/{}/oauth2/v2.0/devicecode", &aad_tenant_id))
        .grant_type("urn:ietf:params:oauth:grant-type:device_code")
        .build()
        .await?;
    let token = authenticator.token(scopes).await?;
```

ggriffiniii commented:
I can fix that in this branch, but just to confirm this isn't a regression with my changes, correct? this exists in the current master as well?

Also do you know how to use https://doc.rust-lang.org/std/os/windows/fs/trait.OpenOptionsExt.html to set the permissions correctly on windows? I don't know windows well.

ctaggart commented:
I don't believe this is a regression. I'm only interested in permissions on Linux & Mac.

ggriffiniii committed:
Create the token file with more secure permissions on unix.

This creates files with 0600 permissions on unix. Still the default
permissions on non-unix platforms.

ggriffiniii commented:
The token file should now get created with more secure permissions on unix. If anybody wants to tell me how to do it on windows I will update it too, but I'm not particularly motivated to learn.

dermesser commented:
just checking by to see how it is going! It sounds like this may take a bit while we're waiting for a new hyper-rustls? Let me know when you want any changes to be merged :)

jroesch commented:
Does anyone know what hyper-rusttls is blocked on? it looks like they changes bumping the version to a point release got merged but there hasn't been a crates release yet.

ggriffiniii commented:
> Does anyone know what hyper-rusttls is blocked on? it looks like they changes bumping the version to a point release got merged but there hasn't been a crates release yet.

I believe it was actually just released about 15 minutes ago. Now everything is just waiting on me to port the changes to hyper 0.13. I was surprised by the number of changes made between the final 0.13 and the prior alpha, but I think the additional changes in 0.13.1 can make it a pretty trivial transition now.

ggriffiniii committed:
Switch from mockito to httptest

ggriffiniii committed:
Move to hyper 0.13.1!!!!

ggriffiniii commented:
Okay, switched to the released version of hyper. I see I'm behind a couple commits on master. I'll rebase over those changes tomorrow, but otherwise I think this is ready for review.

ggriffiniii commented:
I've rebased the PR on top of the latest master. It turns out the issue addressed by the latest commits was already handled as a byproduct of my refactor so there weren't any changes in the end result.

Just to restate what I said earlier. This is now a PR of 71 commits which I realize is rather huge. While it started as a mostly mechanical change, many of the later changes were much more opinionated and I would welcome some discussion on the pros and cons of each.

dermesser commented:
I realize it is quite opinionated, but I see some other users here as
well. You all probably use this library more in real life than I do, so
I tend to trust your changes. For a more fundamental discussion I will
need to take a closer look first, which I hope is ok.
> [
> {
> "@context": "http://schema.org",
> "@type": "EmailMessage",
> "potentialAction": {
> "@type": "ViewAction",
> "target": "
> https://github.com/dermesser/yup-oauth2/pull/112?email_source=notifications\u0026email_token=AAH6PFZMVTJXJ34EP7RJVYDQZJL37A5CNFSM4JLA55HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHG3L3A#issuecomment-567129580
> ",
> "url": "
> https://github.com/dermesser/yup-oauth2/pull/112?email_source=notifications\u0026email_token=AAH6PFZMVTJXJ34EP7RJVYDQZJL37A5CNFSM4JLA55HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHG3L3A#issuecomment-567129580
> ",
> "name": "View Pull Request"
> },
> "description": "View this Pull Request on GitHub",
> "publisher": {
> "@type": "Organization",
> "name": "GitHub",
> "url": "https://github.com"
> }
> }
> ]

ctaggart commented:
Happy new year! Any chance that we can get this merged to master in January?

dermesser commented:
yes, we can! I will probably not have the time to review this very deeply, but it looks like you've done some great work. Do you think I should merge this as is?

ctaggart commented:
@dermesser, I think you should merge it as is. @ggriffiniii  has done some great work. What do you think about adding him as a contributor and letting him merge it?

dermesser commented:
> @dermesser, I think you should merge it as is. @ggriffiniii has done some great work. What do you think about adding him as a contributor and letting him merge it?

Good point! @ggriffiniii I added you as collaborator, meaning you should have push access now. This also means you can continue without being blocked on my response.

I will publish it to crates.io as version 4.0.0 once this is merged.

obiesie commented:
@ggriffiniii great work here!! Any reason why you have opted to go with package private scope for some of the structs like Storage? It would make it more easier to extend if these were public

ggriffiniii commented:
> @ggriffiniii great work here!! Any reason why you have opted to go with package private scope for some of the structs like Storage? It would make it more easier to extend if these were public

There were a few factors for why I opted for an enum rather than keeping the Storage trait.

* Traits that return futures require allocation and virtual dispatch. This seemed unfortunate for a caching layer that frequently will not require any I/O.
* I couldn't find anyway with the old API for users to actually provide their own Storage implementation. The builders didn't provide a way to provide your own storage it only gave the choice between in-memory or on-disk.

Given that Storage appeared to only be used for internal purposes an enum seemed like a better fit that also avoids requiring allocation for the future.

That said I would like to hear how making them public would make it easier to extend. Could you elaborate on your goals?

obiesie commented:
I am looking into adding some code that can handle the google auth flow and from first looks it seems the privacy policy will make this a bit harder.

ggriffiniii commented:
> I am looking into adding some code that can handle the google auth flow and from first looks it seems the privacy policy will make this a bit harder.

Sorry, I was asking for an elaboration to see how I might be able to change the code to address your needs. This description doesn't really explain what you are trying to do with enough detail to be helpful to the discussion I'm afraid. I'm open to adjusting things if it improves the user experience, but I would need to know how the current implementation falls short. It's not just that the types have private visibility. Making `Storage` publicly accessible would not allow you to extend it, nor would it allow you to hand a different implementation to the `Authenticator` without other changes so it's not very clear why you want it to be made visible.

I'm also not quite sure what "google auth flow" would mean in this context other than using the installed flow, device flow, or service account to get oauth2 tokens from google servers, which this crate should already support.

Again I'm open to making changes to address shortcomings, but I need you to elaborate on what you're trying to do so we can try and come up with a way to meet your goals.

ernieturner commented:
FWIW, I just pointed my application to your fork @ggriffiniii and was able to get everything working after a little bit of rewriting. So I'd also vote to have this get merged as well. It seems like the Storage request should probably be done as a separate PR at this point. There's probably enough changes in this PR as is and I'm not sure adding to it is going to help.

Thanks so much for doing the work on this @ggriffiniii, this was one of the last libraries I've been waiting on to get moved over to async.

ggriffiniii commented:
Ok. I pushed the changes into master.

@dermesser: I would suggest creating a 4.0.0-alpha.1 release before the real 4.0.0. Hopefully, that will get more eyes on the changes in case there's any changes that should happen before committing to the public api.

ggriffiniii commented:
@dermesser: Any estimate when we may see a 4.0.0-alpha.1 released on crates.io?

ggriffiniii commented:
> @dermesser: Any estimate when we may see a 4.0.0-alpha.1 released on crates.io?
Nevermind. I see it was released a week ago. Don't know how I missed that. Thanks!

dermesser commented:
don't worry :) I also just noticed I had forgotten to push the version bump commit.

dermesser commented:
> Ok. I pushed the changes into master.
> 
> @dermesser: I would suggest creating a 4.0.0-alpha.1 release before the real 4.0.0. Hopefully, that will get more eyes on the changes in case there's any changes that should happen before committing to the public api.

about this - I think I missed your message about this, but I am not worrying here. It's cheap to release this branch as 4.x and upgrade to 5.x later if it turns out there's something bad about it.
